File: app/src/main/java/com/ryanrvldo/moviecatalogue/data/DataSource.java
package com.ryanrvldo.moviecatalogue.data;

import androidx.lifecycle.LiveData;

import com.ryanrvldo.moviecatalogue.data.source.model.Movie;
import com.ryanrvldo.moviecatalogue.data.source.model.Search;
import com.ryanrvldo.moviecatalogue.data.source.model.Season;
import com.ryanrvldo.moviecatalogue.data.source.model.TvShow;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.MovieResponse;
import com.ryanrvldo.moviecatalogue.data.source.remote.response.TvShowResponse;

import java.util.List;

public interface DataSource {
    LiveData<MovieResponse> getMovies(String category);

    LiveData<MovieResponse> searchMovies(String query);

    LiveData<MovieResponse> getNewReleaseMovies();

    LiveData<Movie> getMovieDetail(int movieId);

    LiveData<TvShowResponse> getTvShows(String category);

    LiveData<TvShowResponse> searchTvShows(String query);

    LiveData<TvShowResponse> getNewReleaseTvShows();

    LiveData<TvShow> getTvDetail(int tvShowId);

    LiveData<Season> getSeasonDetail(int tvId, int seasonNumber);

    LiveData<List<Movie>> getFavMovies();

    void insertFavMovie(Movie movie);

    void deleteFavMovie(Movie movie);

    Movie getFavMovie(int movieId);

    LiveData<List<TvShow>> getFavTvShows();

    void insertFavTv(TvShow tvShow);

    void deleteFavTv(TvShow tvShow);

    TvShow getFavTv(int tvShowId);

    LiveData<List<Search>> getSearchHistory();

    void insertSearch(Search search);

    void deleteSearch(Search search);

    void deleteAllSearchHistory();

    String getSearch(String query);
}

File: app/src/main/java/com/ryanrvldo/moviecatalogue/data/DataSource.kt
package com.ryanrvldo.moviecatalogue.data

import androidx.lifecycle.LiveData
import com.ryanrvldo.moviecatalogue.data.model.Movie
import com.ryanrvldo.moviecatalogue.data.model.Search
import com.ryanrvldo.moviecatalogue.data.model.Season
import com.ryanrvldo.moviecatalogue.data.model.TvShow
import com.ryanrvldo.moviecatalogue.data.remote.response.MovieResponse
import com.ryanrvldo.moviecatalogue.data.remote.response.TvShowResponse
import com.ryanrvldo.moviecatalogue.data.vo.Resource

interface DataSource {
    fun getMovies(category: String): LiveData<Resource<MovieResponse>>

    fun searchMovies(query: String): LiveData<Resource<MovieResponse>>

    fun getNewReleaseMovies(): LiveData<MovieResponse>

    fun getMovieDetails(movieId: Int): LiveData<Resource<Movie>>

    fun getTvShows(category: String): LiveData<Resource<TvShowResponse>>

    fun searchTvShows(query: String): LiveData<Resource<TvShowResponse>>

    fun getNewReleaseTvShows(): LiveData<TvShowResponse>

    fun getTvShowDetails(tvShowId: Int): LiveData<Resource<TvShow>>

    fun getSeasonDetails(tvId: Int, seasonNumber: Int): LiveData<Season>

    fun getFavoriteMovies(): LiveData<List<Movie>>

    suspend fun addFavoriteMovie(movie: Movie)

    suspend fun removeFavoriteMovie(movie: Movie)

    suspend fun isFavoriteMovie(movieId: Int): Boolean

    fun getFavoriteTvShows(): LiveData<List<TvShow>>

    suspend fun addFavoriteTvShow(tvShow: TvShow)

    suspend fun removeFavoriteTvShow(tvShow: TvShow)

    suspend fun isFavoriteTvShow(tvShowId: Int): Boolean

    fun getSearchHistories(): LiveData<List<Search>>

    suspend fun addSearchQuery(search: Search)

    suspend fun removeSearchQuery(search: Search)

    suspend fun removeSearchHistories()

    suspend fun selectSearchQuery(query: String): String?
}